sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ShufflePartitionsUtil.scala

```diff
@@ -117,6 +118,112 @@ object ShufflePartitionsUtil extends Logging {
       return Seq.empty
     }
 
+    val emptyIndexSet = collection.mutable.Set.empty[Int]
+    inputPartitionSpecs.foreach(_.get.iterator.zipWithIndex.foreach {
+      case (EmptyPartitionSpec, i) => emptyIndexSet.add(i)
+      case _ =>
+    })
+
+    if (emptyIndexSet.isEmpty) {
+      return coalescePartitionsWithSkew(mapOutputStatistics, inputPartitionSpecs,
+        targetSize, minPartitionSize, true)
+    }
+
+    // when all partitions are empty, return single EmptyPartitionSpec here to satisfy
+    // SPARK-32083 (AQE coalesce should at least return one partition).
+    if (inputPartitionSpecs.flatten.flatten.forall(_ == EmptyPartitionSpec)) {
+      return inputPartitionSpecs.map(_ => Seq(EmptyPartitionSpec))
+    }
+
+    // ShufflePartitionSpecs at these emptyIndices can NOT be coalesced
+    // split inputPartitionSpecs into sub-sequences by the empty indices, and
+    // call coalescePartitionsWithSkew to optimize each sub-sequence.
+    // let inputPartitionSpecs are:
+    //   [A0(empty), A1, A2, A3(empty), A4(empty), A5, A6, A7, A8, A9, A10]
+    //   [B0, B1, B2, B3, B4(empty), B5, B6, B7, B8(empty), B9, B10]
+    // then:
+    // 1, specs at index (0, 3, 8) are kept: (A0(empty)-B0), (A3(empty)-B3), (A8-B8(empty))
+    // 2, specs at index 4 are discarded, since they are all empty: (A4(empty)-B4(empty))
+    // 3, sub-sequences [A1-B1, A2-B2], [A5-B5, A6-B6, A7-B7], [A9-B9, A10-B10] are optimized
```

Comment:
Would this be a regression ?

If the advisoryPartitionSizeInBytes size is 60 and the two sides of 4 partition size are:
[10, 10, 0, 10]
[10, 0, 10, 10]

The `CoalesceShufflePartitions` will return one reduce task. But if we mark it as empty, we will get three reduce tasks.

Reply:
Yes, this maybe a regression.

This rule may skiped the second and third partitions (if the jointype is `Inner`), but it will split one task into more tasks.

if partitions can be coalesed by `CoalesceShufflePartitions`, we may need to measure whether it is worthwhile to apply this rule (maybe we only mark a partition empty when its size is larger than a threshold)

if  `CoalesceShufflePartitions` is not triggered, it should be always beneficial to mark some non-empty partitions empty.

Reply:
A simple way may be: we can put this rule after the `CoalesceShufflePartitions`, so if we match a empty partition that means we can always skip it for no harm. But compared with applying the rule before `CoalesceShufflePartitions`, we may miss to optimize some empty partition since they may be coalesced.

But I think it seems can be accepted. If the empty partition can be coalesced with other partition, that said the corresponding partition of the other side of join is also small. So if we prefer to coalesce, we actually have small 
overhead since the shuffle data is small enough (less than advisory size). Or even faster than skip empty partition since we have less reduce partition number.

So what we really want to skip is one side partition is empty and other side is big (may be a little skew or can not be optimized through skew). What do you thank ?

Reply:
I think it is nice to put this rule after `CoalesceShufflePartitions`.
To make sure no regression.